### PR TITLE
[MRG] callbacks and modified convergence criteria

### LINKED ISSUE #232: Negative pseudo-R2 on the training set
First off, thanks all for the amazing work that has gone into this package. I'm really excited to use this to analyze my data, it fills a big hole in the existing resources that are out there.

I'm using this to fit a Poisson GLM to explain spike counts using a design matrix with 750K rows and ~100 feature columns. A lot of the features are dummy binary indicators for categorical variables. For demonstration purposes, alpha=0 and reg_lambda=0. I get a negative pseudo-R2 (about -.04) when I score the model on the training set (the same data that was used to fit the model). Probably I'm just doing something stupid but I don't understand how this is even possible. It's my understanding that, at a minimum, the model should be able to predict the mean firing rate and achieve a pseudo-R2 of zero.

Here's my code:
```
glm = pyglmnet.GLM(distr='poisson', alpha=0, reg_lambda=[0], score_metric='pseudo_R2', verbose=True)
glm.fit(exog.values, endog.values)
```

This prints:

> Looping through the regularization path
> Lambda: 0.0000
> 	Converged. Loss function: 0.04
> 	dL/L: -0.000996

The score is negative on the training set:

> In [13]: glm.score(exog.values, endog.values)
> Out[13]: array([-0.04439965])

What can I try to debug this?

I would be happy to provide the data if it would be helpful but the matrices are quite large so let me know if that is necessary. `exog` is a pandas DataFrame with zero mean and unit variance. `endog` is a pandas Series of spike counts (4318 spikes total, mean .006 spikes per bin).

### PR DESCRIPTION
**Pull request by pavanramkumar**
replaces #233 anf closes #232

### CODE REVIEW

**Comment by pavanramkumar** [comment]
`test_glmcv()` takes very long to pass and i think it's because of repeated fits in each fold for each `reg_lambda`. 

i'm not yet convinced that we're getting the `GLMCV()` use case right with respect to what happens in R's glmnet, i.e. just fit the model along a regularization path with no cross-validation. i want to study the scikit-learn `LassoCV` use case more closely before we merge this.

File: pyglmnet/pyglmnet.py [hunk 1]
```diff
@@ -446,6 +446,8 @@ class GLM(BaseEstimator):
     random_state : int
         seed of the random number generator used to initialize the solution.
         default: 0
+    callbacks: boolean
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: why not make this a `callable` (instead of bool) which can be any function and is called at every iteration. If it's `None` nothing happens. So, you embed all the logic inside this callback

Comment (jasmainak) [reply pyglmnet/pyglmnet.py]: a bit like this: https://github.com/alphacsc/alphacsc/blob/a87b48cd7229af4c5cee4f94a334e2790d21e750/alphacsc/learn_d_z.py#L84

**Comment by pavanramkumar** [comment]
Ok, i reviewed `LassoCV` again and it does seem equivalent to `GLMCV` in what it does, which is in-turn equivalent to R's `cv.glmnet`.

However, we don't have a simple class that fits on a regularization path without selecting for the best `reg_lambda` with cross-validation. Something to think about for the future.

LGTM

File: pyglmnet/pyglmnet.py (hunk 1 above)
Comment (pavanramkumar) [reply pyglmnet/pyglmnet.py]: nice suggestion, done!

**Comment by jasmainak** [comment]
@pavanramkumar :

> However, we don't have a simple class that fits on a regularization path without selecting for the best reg_lambda with cross-validation. Something to think about for the future.

I would argue that you can give `cv=1` and that would do it for you, no?

I removed `callback` from the public API (that is no documentation) for now because we use it only for testing and there is no compelling use case for users. Is it okay for you?

Also, I think we should document the stopping criteria. +1 for MRG once this is done!

**Comment by pavanramkumar** [comment]
@jasmaink looks good, it's fine to remove `callback` from the public API. for documenting convergence, i'd just add a comment line; something like:

```
# convergence achieved if learning_rate * norm(gradient) < tolerance
```

can you add this?

File: pyglmnet/pyglmnet.py [hunk 2]
```diff
@@ -714,20 +716,30 @@ def fit(self, X, y):
...
+        # Iterative updates
         for t in range(0, self.max_iter):
             if self.solver == 'batch-gradient':
                 grad = _grad_L2loss(self.distr,
                                     alpha, self.Tau,
                                     reg_lambda, X, y, self.eta,
                                     beta)
-
+                if t > 1:
+                    if np.linalg.norm(grad) / np.linalg.norm(beta) < tol / lr:
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: @pavanramkumar couldn't this be done using a single `if` statement? According to zen of python:

> Flat is better than nested

File: pyglmnet/pyglmnet.py [hunk 3]
```diff
@@ -714,20 +716,30 @@ def fit(self, X, y):
...
                 beta = beta - self.learning_rate * grad
+
             elif self.solver == 'cdfast':
+                beta_old = deepcopy(beta)
                 beta, z = \
                     self._cdfast(X, y, z, ActiveSet, beta, reg_lambda)
-
+                if t > 1:
+                    if ((np.linalg.norm(beta - beta_old) /
+                         np.linalg.norm(beta_old) < tol / lr)):
```
Comment (jasmainak) [inline pyglmnet/pyglmnet.py]: The denominator here is `norm(beta_old)` but in the other solver it is `norm(beta)`. Is that expected?

**Comment by jasmainak** [comment]
> # convergence achieved if learning_rate * norm(gradient) < tolerance

I also noticed a `norm(beta)` in the denominator ... can you clarify why that is needed?

**Comment by jasmainak** [comment]
@pavanramkumar just noticed a couple more things. Let me know. I can fix them for you if needed

File: pyglmnet/pyglmnet.py (hunk 2 above)
Comment (pavanramkumar) [reply pyglmnet/pyglmnet.py]: hmm, we get computational savings by not having to compute gradients and parameter norms in the first iteration

Comment (jasmainak) [reply pyglmnet/pyglmnet.py]: see https://github.com/glm-tools/pyglmnet/pull/252#issuecomment-419670231

**Comment by pavanramkumar** [comment]
@jasmainak made all the changes we discussed. merge?

**Comment by jasmainak** [comment]
@pavanramkumar you now have a merge commit ... you can do interactive rebase to remove it. Let me know if I can push, then I can do it for you.

**Comment by jasmainak** [comment]
Otherwise LGTM

**Comment by pavanramkumar** [comment]
@jasmainak yes the merge commit didn't catch all the conflicts, so the tests failed. you can go ahead and rebase it!

**Comment by jasmainak** [comment]
okay rebased. Let's hope tests pass now

**Comment by pavanramkumar** [comment]
fixed conflicts. merging

**Comment by jasmainak** [comment]
@pavanramkumar looks like it slowed down one of the examples by quite a bit so CircleCI is now failing ...
